### PR TITLE
work with JMAP::Tester futures branch

### PR DESCRIPTION
**Pull request by rjbs**
See https://github.com/fastmail/JMAP-Tester/pull/43

### CODE REVIEW

File: lib/Ix/DBIC/Result.pm [hunk 1]
```diff
@@ -329,10 +329,6 @@ sub ix_update_extra_search ($self, $ctx, $arg) {
   );
 }
 
-sub ix_update_extra_select {
```
Comment (wolfsage) [inline lib/Ix/DBIC/Result.pm]: I think you're including two unrelated commits in this mr?

File: t/access_log.t [hunk 2]
```diff
@@ -46,7 +50,8 @@ my %common = (
 );
 
 # Should be ignored, behind_proxy defaults to disabled.
-$jmap_tester->ua->default_header('X-Forwarded-For' => '1.2.3.4');
+# XXX This is assuming an LWP UA. -- rjbs, 2019-04-24
+$jmap_tester->ua->lwp->default_header('X-Forwarded-For' => '1.2.3.4');
```
Comment (wolfsage) [inline t/access_log.t]: Yeah... we will probably want to proxy some of the methods we expect folks to use on the ua - light mucking with cookies and headers...

**Comment by wolfsage** [comment]
Seems right

File: lib/Ix/DBIC/Result.pm (hunk 1 above)
Comment (rjbs) [reply lib/Ix/DBIC/Result.pm]: Removed.

**Comment by wolfsage** [comment]
Still seems right

**Comment by wolfsage** [comment]
Oh, dist.ini we'll want to update the JMAP::Tester prereq to 0.027 assuming the Futures release is that version number.

File: t/access_log.t [hunk 3]
```diff
@@ -17,7 +17,11 @@ use Test::More;
 my ($app, $jmap_tester) = Bakesale::Test->new_test_app_and_tester;
 \my %account = Bakesale::Test->load_trivial_account($app->processor->schema_connection);
 
-$jmap_tester->_set_cookie('bakesaleUserId', $account{users}{rjbs});
```
Comment (mmcclimon) [inline t/access_log.t:20]: As part of my "why not make upgrading as easy as possible" campaign, it should be easy to shim this in JMAP::Tester so that `_set_cookie` continued to do the right thing, so that no change would be needed here.

File: t/basic.t [hunk 4]
```diff
@@ -2481,7 +2499,14 @@ subtest "exeptions are not thrown twice" => sub {
 
   $app->processor->clear_exceptions;
 
-  my (undef, $res) = capture_stderr(sub { $jmap_tester->ua->get($uri) });
+  my (undef, $res) = capture_stderr(sub {
+    $jmap_tester->ua->request(
+      $jmap_tester,
+      HTTP::Request->new(GET => $uri),
+      'jmap',
```
Comment (mmcclimon) [inline t/basic.t]: This passing 'jmap' is a smell, to me:

a) 'jmap' is effectively just a magic string
b) This one in particular isn't even a JMAP request, so knowing that you need to pass 'jmap' here is relying on the guts of the JMAP::Tester logger object.

At the very least, it feels like the logger should have some generic fallback, if one of the magic strings isn't present. Also, 'jmap' as a string is weird in the first place, since the thing is on a JMAP::Tester object. ('upload' and 'download' are also part of JMAP!)

Comment (mmcclimon) [reply t/basic.t]: (Update, after reading some JMAP::Tester code: it looks like `jmap` is the default for this, so you could probably just eliminate 'jmap' here, but I think the rest of my comment is still relevant.)

File: t/basic.t [hunk 5]
```diff
@@ -2481,7 +2499,14 @@ subtest "exeptions are not thrown twice" => sub {
...
+    )->get;
```
Comment (mmcclimon) [inline t/basic.t]: Hm. This `->get` also feels like a smell to me. That's because: in general, if you just have a JMAP::Tester, you don't need to _know_ that it traffics in Futures internally, and only if you're doing stuff with its UA directly.

It feels to me like the UA should inherit the Tester's `should_return_futures` setting (which would be possible if there were, as suggested elsewhere, a weak ref back to the tester from the UA). Then `$tester->ua->request` would just do the right thing, based on how you'd configured the tester, and you wouldn't have this telltale `->get`. (I know that it isn't _really_ an implementation detail, because `->request` is doecumented to return a Future, but it still _feels_ like one, IMO.)

File: t/access_log.t (hunk 2 above)
Comment (mmcclimon) [reply t/access_log.t]: This should now use `set_default_header` and not call `->lwp`.

File: t/access_log.t [hunk 6]
```diff
@@ -159,10 +164,15 @@ for my $line (@lines) {
...
+  $jmap_tester->ua->set_cookie({
+    api_uri => $jmap_tester->api_uri,
+    name    => 'bakesaleUserId',
+    value   => $account{users}{rjbs},
+  });
 
   # Our real request ip!
-  $jmap_tester->ua->default_header('X-Forwarded-For' => '1.2.3.4');
+  # XXX This is assuming an LWP UA. -- rjbs, 2019-04-24
+  $jmap_tester->ua->lwp->default_header('X-Forwarded-For' => '1.2.3.4');
```
Comment (mmcclimon) [inline t/access_log.t]: This should use `set_default_header`.

**Review by mmcclimon — COMMENTED** [review]
I have left some comments, mostly having to do with changes I'd like to see in JMAP::Tester itself rather than in the particulars of this code (though I left a few of those too).

Given my comments in JMAP::Tester and also here, I'm not convinced it's going to be super useful for me to review all of the other "cope with futures branch" changes right this second, since a lot of my suggestions are "I think we should make these coping strategies unnecessary."

I could maybe be convinced otherwise, but I think I'll stop for today!

File: t/access_log.t (hunk 2 above)
Comment (rjbs) [reply t/access_log.t]: I did the heck out of both those requests.

File: t/basic.t (hunk 4 above)
Comment (rjbs) [reply t/basic.t]: This goes away with the new http_X methods.

File: t/basic.t (hunk 5 above)
Comment (rjbs) [reply t/basic.t]: This has been resolved with the http_X methods.

File: t/access_log.t (hunk 6 above)
Comment (rjbs) [reply t/access_log.t]: Done.

**Review by mmcclimon — APPROVED** [review]
Seems right, and the new JMAP::Tester abstractions (esp. `http_*`) are clearly better.
